### DISCUSSION TITLE
Typed throws (why not)

### DISCUSSION THREAD

**Discussion by dabrahams** [hylo-lang/hylo#1257] [Language design]
This is a place to point people to when statically specifying what errors can be thrown from a function comes up for discussion.

I oppose such a feature for the following reasons.  That isn't to say we won't entertain counter-arguments, but they will have to outweigh these reasons and the added language complexity will need to seem justified.

- Throwing is optimal for handling conditions far from their source (use `Result`, etc. for the lowest-level operations that might need to be retried, or where other error-specific recovery actions are likely)
- Far from the source of the error, there is rarely anything you can do differently to recover based on the details of the error.  There is usually some general procedure, like "roll back any changes, report the failure to the user, and continue"
- That is to say, speaking broadly, code does not care about the types of thrown errors: the information they contain is useful only for later analysis by humans.  Make sure there is a way to get a useful textual representation.
- Decorating code with information about which errors can propagate therefore adds a lot of syntax in order to control something statically that nobody cares about statically.
- Programmers love to take advantage of anything that gives them a sense of control so they will use it.
- Programmers who want it (including the few that need it) will demand it of library authors so it will proliferate.

**Comment by waydan** [comment]
This is a very nice summary of concerns regarding statically specifying error types. I was thinking of responding to the ongoing design of #1261, but I think feedback is more appropriate here. I'm coming from a real-time embedded domain, so my needs may differ from others.
>Throwing is optimal for handling conditions far from their source (use `Result`, etc. for the lowest-level operations that might need to be retried, or where other error-specific recovery actions are likely)

To me, throwing could be optimal for expressing local errors two since I can separate a normal path and an error path cleanly in code. I think language "throw" feature is superior to a `Result` type for the following reasons.
1. I think ABI could be designed to take advantage of a language feature more easily than a specific library type since the latter would require a special optimization rule.
2. If a `Result` type was really needed (e.g., using a monadic `.and_then()`), an library function or special syntax could easily support this functionality.

>That is to say, speaking broadly, code does not care about the types of thrown errors: the information they contain is useful only for later analysis by humans. Make sure there is a way to get a useful textual representation.

I may be reading into this, but "code does not care about the types of thrown errors" sounds like type-erasure. I want exceptions without dynamic allocation, but I don't know how to return exceptions on the stack unless the compiler knows the type. I won't throw strings, vectors, or other types with unbounded size, but maybe I'm going to throw an enum or struct which could be logged.

>Programmers who want it (including the few that need it) will demand it of library authors so it will proliferate.

This is kind of a problem in C++ where we have throwing, non-throwing, and functions that signal errors another way. I would prefer if there was "one true way" to signal an error that was useable by all and all library authors could be reasonably expected to adhere to. I recently watched [Peter Muldoon's C++Now talk](https://www.youtube.com/watch?v=HXJmrMnnDYQ) in which he highlights how difficult it is for a library author to anticipate how their code will be applied by a user. To me, offering a `throws` and a `Result` feeds that burden. If `throws` could be converted by users who need it and was also efficient/manageable for even constrained systems, then it would rise as the preferred method for all error propagation. Peter's thesis seems to be that exceptions (in C++) are expensive, so they should only be used for reporting near the top of an executable. If they were more efficient, I think we could use them elsewhere.

I definitely appreciate the burden of adding an error type as a generic parameter or spelling it explicitly. If there was a way to avoid this without loosing the ability to use the stack for exceptions I would support it. I'm not sure how that can be achieved, but that's due to my insufficient knowledge.

**Reply by dabrahams** [reply]
> I can separate a normal path and an error path cleanly in code.

How does  `if errorOccurred { ... } else { ... }` (which doesn't depend on throwing) fail to separate normal paths from error paths?

> I think language "throw" feature is superior to a Result type…

Correct me if I'm wrong, but I think which one is superior is irrelevant, because Hylo will have "throw"

> "code does not care about the types of thrown errors" sounds like type-erasure. 

Yes, it is exactly type erasure.

> I want exceptions without dynamic allocation, but I don't know how to return exceptions on the stack unless the compiler knows the type. I won't throw strings, vectors, or other types with unbounded size, but maybe I'm going to throw an enum or struct which could be logged.

I'm pretty sure you want throw without the usual _costs_ of dynamic allocation (even stack allocations are dynamic after all). There are ways we can minimize the cost required to propagate errors of bounded size, and since our existential types have an inline buffer, if your thrown objects are small enough there will be zero dynamic allocation cost. 

The fact that it is difficult for a library author (or language designer) to anticipate how their code will be applied doesn't let us off the hook for anything.  Knowing (and hypothesizing) about how our products will be used is part of the job; we have to design to make the important cases easy, elegant, safe, and efficient, and ideally to make the other cases possible.  I've been studying effective usage patterns for exceptions, and how various language design choices play out, informed by real code, since 1997.  So while I may not be able to anticipate everything that's coming, my hypotheses have a pretty decent chance of holding up.

>  I would prefer if there was "one true way" to signal an error that was useable by all and all library authors could be reasonably expected to adhere to.

Sure, “one way to do things” is a nice principle in isolation, but every language design choice has downstream consequences.  When even errors to be handled by the caller are reported as exceptions, people reasonably want statically typed throws, which leads to the costs outlined at the top (not to mention hurting readability in the caller).  As I wrote above the advantages of upholding this nice principle  would have to outweigh many costs including that of increased language complexity.  So far the advantage seems to be that the guideline for how to report errors is slightly simpler.  Did I miss anything?

I don't know if this will make you feel better, but: aside from a trivial syntax transformation, the only difference between statically-typed throws and a result type is that an exception propagates through the stack automatically.  But automatic propagation is counterproductive when the condition is to be handled by the immediate caller; they need to actively defeat propagation.  The feature is a bad match for the use case.  And this use case is extremely rare: it's mostly the very lowest-level APIs (e.g. try once to send a packet) that will be wrapped in something that _does_ throw, so your world will still be relatively uniform.
